src/main.cpp

```diff
@@ -6355,7 +6355,25 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
         bool fCanDirectFetch = CanDirectFetch(chainparams.GetConsensus());
         CNodeState *nodestate = State(pfrom->GetId());
-        nodestate->fFirstHeadersReceived = true;
+
+        // During the initial peer handshake we must receive the initial headers which should be greater
+        // than or equal to our block height at the time of requesting GETHEADERS. This is because the peer has
+        // advertised a height >= to our own. Furthermore, because the headers max returned is as much as 2000 this
+        // could not be a mainnet re-org.
+        if (!nodestate->fFirstHeadersReceived)
+        {
+            // We want to make sure that the peer doesn't just send us any old valid header. The last header block
+            // hash they send us should now be equal to our current pindexLast block hash.
+            if (hashLastBlock == pindexLast->GetBlockHash())
+            {
```

Comment:
if they send an old header then they will not be equal...hash last block will be the hash of the last header in the list they sent....whereas pindexLast will only update if the header advances the chain.   In general they will be sending the header to the chain tip, so they will still be equal.

yes, i have to put in a check for pindexLast.